src/storage/src/storage_state.rs

```diff
@@ -899,15 +903,51 @@ impl<'w, A: Allocate> Worker<'w, A> {
             .keys()
             .collect::<BTreeSet<_>>();
         let mut stale_exports = self.storage_state.exports.keys().collect::<BTreeSet<_>>();
+
+        // First, we collect all "drop commands". These are `AllowCompaction`
+        // commands that compact to the empty since. Then, later, we make sure
+        // we retain only those `Create*` commands that are not dropped. We
+        // assume that the `AllowCompaction` command is ordered after the
+        // `Create*` commands but don't assert that.
+        // WIP: Should we assert?
+        let mut drop_commands = BTreeSet::new();
+
+        for command in &mut commands {
+            match command {
+                StorageCommand::CreateTimely { .. } => {
+                    panic!("CreateTimely must be captured before")
+                }
+                StorageCommand::AllowCompaction(sinces) => {
+                    for (id, since) in sinces.iter() {
```

Comment:
You can use `drain_filter_swapping` from `mz_ore`:

```rust
let drops = sinces.drain_filter_swapping(|(_id, since)| since.is_empty());
drop_commands.extend(drops.map(|(id, _since)| *id));
```

Reply:
nice! I didn't know that one.